File: plugins/plugin-imessage/rust/src/actions/send_message.rs
//! Send message action for iMessage

use crate::service::IMessageService;
use crate::types::{format_phone_number, is_email, is_phone_number};
use crate::IMESSAGE_SERVICE_NAME;
use async_trait::async_trait;
use elizaos::{Action, ActionExample, Content, IAgentRuntime, Memory, State};
use regex::Regex;
use tracing::{debug, error, info, warn};

/// Action to send a message via iMessage
pub struct SendMessageAction;

impl SendMessageAction {
    /// Creates a new send message action
    pub fn new() -> Self {
        Self
    }
}

impl Default for SendMessageAction {
    fn default() -> Self {
        Self::new()
    }
}

/// Extract a phone number or email address from message text.
///
/// Scans the text for patterns that look like phone numbers or email addresses
/// and returns the first valid match, formatted appropriately.
pub fn extract_target_from_text(text: &str) -> Option<String> {
    // Try to find phone numbers (patterns like +1234567890, (123) 456-7890, etc.)
    let phone_re = Regex::new(r"[+]?\d[\d\s\-().]{8,14}\d").unwrap();
    for m in phone_re.find_iter(text) {
        let candidate = m.as_str();
        if is_phone_number(candidate) {
            return Some(format_phone_number(candidate));
        }
    }

    // Try to find email addresses
    let email_re = Regex::new(r"[^\s@]+@[^\s@]+\.[^\s@]+").unwrap();
    for m in email_re.find_iter(text) {
        let candidate = m.as_str();
        if is_email(candidate) {
            return Some(candidate.to_string());
        }
    }

    None
}

#[async_trait]
impl Action for SendMessageAction {
    fn name(&self) -> &str {
        "SEND_IMESSAGE"
    }

    fn description(&self) -> &str {
        "Send a message via iMessage (macOS only)"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "TEXT_MESSAGE",
            "IMESSAGE_SEND",
            "APPLE_MESSAGE",
            "MESSAGE_CONTACT",
        ]
    }

    fn examples(&self) -> Vec<Vec<ActionExample>> {
        vec![
            vec![
                ActionExample {
                    name: "{{user1}}".to_string(),
                    content: Content {
                        text: Some(
                            "Can you send a message to John saying I'll be late?".to_string(),
                        ),
                        ..Default::default()
                    },
                },
                ActionExample {
                    name: "{{agentName}}".to_string(),
                    content: Content {
                        text: Some("I'll send that message to John for you.".to_string()),
                        action: Some("SEND_IMESSAGE".to_string()),
                        ..Default::default()
                    },
                },
            ],
            vec![
                ActionExample {
                    name: "{{user1}}".to_string(),
                    content: Content {
                        text: Some("Text mom that I'm on my way home".to_string()),
                        ..Default::default()
                    },
                },
                ActionExample {
                    name: "{{agentName}}".to_string(),
                    content: Content {
                        text: Some("I'll send that text to mom now.".to_string()),
                        action: Some("SEND_IMESSAGE".to_string()),
                        ..Default::default()
                    },
                },
            ],
        ]
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        runtime
            .get_service::<IMessageService>(IMESSAGE_SERVICE_NAME)
            .is_some()
    }

    async fn handler(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
        _options: Option<&serde_json::Value>,
    ) -> elizaos::Result<Option<Content>> {
        let service = runtime
            .get_service::<IMessageService>(IMESSAGE_SERVICE_NAME)
            .ok_or_else(|| {
                elizaos::Error::ServiceError("iMessage service not available".to_string())
            })?;

        let text = message
            .content
            .text
            .as_deref()
            .ok_or_else(|| elizaos::Error::ValidationError("No message text".to_string()))?;

        if text.trim().is_empty() {
            warn!("Empty message text, skipping send");
            return Ok(None);
        }

        // Try to extract target (phone number or email) from message content,
        // matching the approach used in TypeScript/Python implementations.
        let extracted_target = extract_target_from_text(text);

        let target = if let Some(ref t) = extracted_target {
            debug!("Extracted iMessage target from text: {}", t);
            t.clone()
        } else {
            // Fall back to room channel_id for reply-in-current-chat
            debug!("No target in text, falling back to room channel_id");
            let room = runtime
                .get_room(&message.room_id)
                .await?
                .ok_or_else(|| elizaos::Error::NotFound("Room not found".to_string()))?;

            room.channel_id.ok_or_else(|| {
                elizaos::Error::ValidationError(
                    "No target found in message and no channel ID for room".to_string(),
                )
            })?
        };

        match service.send_message(&target, text, None).await {
            Ok(result) => {
                if result.success {
                    info!("Sent iMessage to {}: {:?}", target, result.message_id);

                    Ok(Some(Content {
                        text: Some(text.to_string()),
                        source: Some("imessage".to_string()),
                        metadata: Some(serde_json::json!({
                            "messageId": result.message_id,
                            "chatId": result.chat_id,
                            "target": target,
                        })),
                        ..Default::default()
                    }))
                } else {
                    error!("Failed to send iMessage: {:?}", result.error);
                    Err(elizaos::Error::ServiceError(
                        result.error.unwrap_or_else(|| "Unknown error".to_string()),
                    ))
                }
            }
            Err(e) => {
                error!("Failed to send iMessage: {}", e);
                Err(elizaos::Error::ServiceError(e.to_string()))
            }
        }
    }
}

File: plugins/plugin-imessage/src/actions/sendMessage.ts
/**
 * Send message action for the iMessage plugin.
 */

import type {
  Action,
  ActionResult,
  HandlerCallback,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { composePromptFromState, logger, ModelType, parseJSONObjectFromText } from "@elizaos/core";
import type { IMessageService } from "../service.js";
import { IMESSAGE_SERVICE_NAME, isValidIMessageTarget, normalizeIMessageTarget } from "../types.js";

const SEND_MESSAGE_TEMPLATE = `# Task: Extract iMessage parameters

Based on the conversation, determine what message to send and to whom.

Recent conversation:
{{recentMessages}}

Extract the following:
1. text: The message content to send
2. to: The recipient (phone number, email, or "current" to reply)

Respond with JSON only, with no prose or fences:
{
  "text": "message to send",
  "to": "phone/email or current"
}
`;

interface SendMessageParams {
  text: string;
  to: string;
}

const MAX_IMESSAGE_TEXT_CHARS = 4_000;
const IMESSAGE_ACTION_TIMEOUT_MS = 30_000;

function truncateActionText(text: string, maxChars: number): string {
  return text.length > maxChars ? `${text.slice(0, maxChars - 3)}...` : text;
}

function parseSendMessageParams(response: string): SendMessageParams | null {
  const parsed = parseJSONObjectFromText(response) as Record<string, unknown> | null;
  if (parsed?.text) {
    return {
      text: truncateActionText(String(parsed.text), MAX_IMESSAGE_TEXT_CHARS),
      to: String(parsed.to || "current"),
    };
  }

  return null;
}

export const sendMessage: Action = {
  name: "IMESSAGE_SEND_MESSAGE",
  similes: ["SEND_IMESSAGE", "IMESSAGE_TEXT", "TEXT_IMESSAGE", "SEND_IMSG"],
  description: "Send a text message via iMessage (macOS only)",
  descriptionCompressed: "Send iMessage (macOS).",
  contexts: ["phone", "messaging", "connectors"],
  contextGate: { anyOf: ["phone", "messaging", "connectors"] },
  roleGate: { minRole: "USER" },
  parameters: [
    {
      name: "text",
      description: "Message text to send.",
      required: false,
      schema: { type: "string" },
    },
    {
      name: "to",
      description: "Phone number, email address, or current conversation.",
      required: false,
      schema: { type: "string", default: "current" },
    },
  ],
  suppressPostActionContinuation: true,

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    // The action is available whenever the iMessage service is registered,
    // regardless of where the inbound message came from. Previously this
    // validate() required `message.content.source === "imessage"`, which
    // meant the agent could only send iMessages if the *user's* incoming
    // message had also come in over iMessage. That broke the common cases:
    // commanding the agent from the dashboard, from /api/chat, from a
    // cron task, or from any other connector. Opening the gate to any
    // source is the only way "tell my agent to text Shaw" works from
    // outside an iMessage conversation.
    //
    // We still require (a) the service to be registered, so we don't
    // pretend to expose a send path the runtime can't fulfil, and (b)
    // some intent signal in the text — a bare "hi" on the wire from a
    // non-imessage source shouldn't flag this action as relevant.
    const imessageService =
      typeof runtime?.getService === "function"
        ? runtime.getService<IMessageService>(IMESSAGE_SERVICE_NAME)
        : null;
    if (!imessageService) return false;

    const source = String(message.content.source ?? "");

    // Messages that arrive over iMessage itself are always eligible —
    // the agent is mid-conversation and every reply is a candidate send.
    if (source === "imessage") return true;

    // For every other source (client_chat, dashboard, api, cron, …) we
    // look for explicit intent: a structured invocation in the content,
    // or keywords that match the Action's purpose.
    const hasStructuredInvocation =
      Boolean(message.content.actions?.includes("IMESSAGE_SEND_MESSAGE")) ||
      (typeof message.content === "object" &&
        typeof (message.content as Record<string, unknown>).to === "string");
    if (hasStructuredInvocation) return true;

    const text = typeof message.content.text === "string" ? message.content.text.trim() : "";
    if (!text) return false;

    // Keyword intent: "text", "send", "message", "imessage", "sms".
    // Any of these in the user's prompt marks the action as candidate;
    // the LLM still decides whether to actually invoke it.
    return /\b(imessage|text|send|message|sms)\b/i.test(text);
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const imessageService = runtime.getService<IMessageService>(IMESSAGE_SERVICE_NAME);

    if (!imessageService?.isConnected()) {
      if (callback) {
        callback({
          text: "iMessage service is not available.",
          source: "imessage",
        });
      }
      return { success: false, error: "iMessage service not available" };
    }

    if (!imessageService.isMacOS()) {
      if (callback) {
        callback({
          text: "iMessage is only available on macOS.",
          source: "imessage",
        });
      }
      return { success: false, error: "iMessage requires macOS" };
    }

    // Compose state if not provided
    const currentState = state ?? (await runtime.composeState(message));

    // Extract parameters using LLM
    const prompt = await composePromptFromState({
      template: SEND_MESSAGE_TEMPLATE,
      state: currentState,
    });

    let msgInfo: SendMessageParams | null = null;

    for (let attempt = 0; attempt < 3; attempt++) {
      const response = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt,
      });

      const parsed = parseSendMessageParams(response);
      if (parsed?.text) {
        msgInfo = parsed;
        break;
      }
    }

    if (!msgInfo?.text) {
      if (callback) {
        callback({
          text: "I couldn't understand what message you want me to send. Please try again.",
          source: "imessage",
        });
      }
      return { success: false, error: "Could not extract message parameters" };
    }
    msgInfo = {
      ...msgInfo,
      text: msgInfo.text.slice(0, MAX_IMESSAGE_TEXT_CHARS),
    };

    // Determine target
    let targetId: string | undefined;

    if (msgInfo.to && msgInfo.to !== "current") {
      const normalized = normalizeIMessageTarget(msgInfo.to);
      if (normalized && isValidIMessageTarget(normalized)) {
        targetId = normalized;
      }
    }

    // Fall back to current chat
    if (!targetId) {
      const stateData = (currentState.data || {}) as Record<string, unknown>;
      targetId = (stateData.chatId as string) || (stateData.handle as string);
    }

    if (!targetId) {
      if (callback) {
        callback({
          text: "I couldn't determine who to send the message to. Please specify a phone number or email.",
          source: "imessage",
        });
      }
      return { success: false, error: "Could not determine recipient" };
    }

    // Send message
    const timeoutMs = IMESSAGE_ACTION_TIMEOUT_MS;
    const result = await imessageService.sendMessage(targetId, msgInfo.text);

    if (!result.success) {
      if (callback) {
        callback({
          text: `Failed to send message: ${result.error}`,
          source: "imessage",
        });
      }
      return { success: false, error: result.error };
    }

    logger.debug(`Sent iMessage to ${targetId}`);

    return {
      success: true,
      data: {
        to: targetId,
        messageId: result.messageId,
        timeoutMs,
        suppressVisibleCallback: true,
        suppressActionResultClipboard: true,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user1}}",
        content: { text: "Send them a message saying 'Hello!'" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll send that message via iMessage.",
          actions: ["IMESSAGE_SEND_MESSAGE"],
        },
      },
    ],
    [
      {
        name: "{{user1}}",
        content: {
          text: "Text +1234567890 saying 'I'll be there in 10 minutes'",
        },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll send that text.",
          actions: ["IMESSAGE_SEND_MESSAGE"],
        },
      },
    ],
  ],
};
